File: src/main/java/com/gjw/common/utils/JwtUtils.java
package com.gjw.common.utils;

import com.gjw.common.enums.SystemCodeEnums;
import com.gjw.common.result.SingleResult;
import io.jsonwebtoken.*;
import org.apache.commons.codec.binary.Base64;

import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;
import java.util.Date;

/**
 * JWT 工具
 */
public class JwtUtils {

    private static final String secretKey = "123456";
    private static final Long survivalTime = 3600L;

    /**
     * 签发JWT
     *
     * @param id      数据ID
     * @param subject 可以是JSON数据 尽可能少
     */
    public static String createJWT(String id, String subject) {
        SignatureAlgorithm signatureAlgorithm = SignatureAlgorithm.HS256;
        SecretKey scrtKey = generalKey(secretKey);
        long expMillis = System.currentTimeMillis() + survivalTime * 1000;
        Date expDate = new Date(expMillis);
        JwtBuilder builder = Jwts.builder()
                .setId(id)
                .setSubject(subject)// 主题
                .setIssuer("go")// 签发者
                .setIssuedAt(new Date())// 签发时间
                .signWith(signatureAlgorithm, scrtKey)// 签名算法以及密匙
                .setExpiration(expDate); // 过期时间
        return builder.compact();
    }

    /**
     * 验证JWT
     */
    public static SingleResult<Claims> validateJWT(String jwtStr) {
        SingleResult<Claims> result = new SingleResult<>();
        try {
            result.setCode(SystemCodeEnums.SUCCESS.getCode());
            result.setMsg(SystemCodeEnums.SUCCESS.getMsg());
            result.setData(parseJWT(jwtStr));
        } catch (ExpiredJwtException e) {
            result.setCode(SystemCodeEnums.JWT_EXPIRED.getCode());
            result.setMsg(SystemCodeEnums.JWT_EXPIRED.getMsg());
        } catch (Exception e) {
            result.setCode(SystemCodeEnums.FAIL.getCode());
            result.setMsg("验签失败");
        }
        return result;
    }

    private static SecretKey generalKey(String secretKey) {
        byte[] encodedKey = Base64.decodeBase64(secretKey);
        return new SecretKeySpec(encodedKey, 0, encodedKey.length, "AES");
    }

    /**
     * 解析JWT字符串
     */
    public static Claims parseJWT(String jwt) {
        return Jwts.parser().setSigningKey(generalKey(secretKey)).parseClaimsJws(jwt).getBody();
    }
}

File: spring-security-server/src/main/java/com/go/security/JwtUtils.kt
package com.go.security

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import java.util.*

/**
 * Jwt工具类
 */
class JwtUtils {
    companion object {
        private const val EXPIRE_TIME: Long = 1000 * 60 * 60 * 24 * 7 // 过期时间 7天
        private const val SECRET: String = "dskfks3hfowifei3*t978" // 秘钥

        /**
         * 生成token
         */
        fun createToken(map: MutableMap<String, Any>): String {
            return Jwts.builder()
                .setClaims(map)
                .setIssuedAt(Date())//创建 JWT 时的时间戳
                .setExpiration(Date(System.currentTimeMillis() + EXPIRE_TIME))
                .signWith(SignatureAlgorithm.HS256, SECRET)// 设置签名
                .compact()
        }

        /**
         * 根据token解析出用户信息
         * @exception io.jsonwebtoken.ExpiredJwtException token过期异常
         */
        fun parseToken(token: String): Claims {
            return Jwts.parser()
                .setSigningKey(SECRET)
                .parseClaimsJws(token)
                .body
        }
    }
}